pkg/build/cleanup.go

```diff
@@ -11,8 +11,16 @@ import (
 // temporary directories created by STI build and it also cleans the temporary
 // Docker images produced by LayeredBuild
 type DefaultCleaner struct {
-	util.FileSystem
-	docker.Docker
+	fs     util.FileSystem
+	docker docker.Docker
```

Comment:
> Thank you, I knew that there is no inheritance, but it still looks similar -- struct has additional methods from another struct.
> 
> > Before this patch, the exported type DefaultCleaner had fields FileSystem and Docker, plus all exported fields of those two latter types
> I didn't know about these 2 fields. Do you mean that I could write like this?
> 
> c.FileSystem.RemoveDirectory(config.WorkingDir)


Yes! Normally the point of embedding is that you don't need the extra dot.

These are equivalent:

```go
// using promoted method
c.RemoveDirectory(config.WorkingDir)

// using explicit method
c.FileSystem.RemoveDirectory(config.WorkingDir)
```


I think the original usage of embedding here was not very idiomatic nor necessary.

One common usage of that is to add a mutex to control access to some struct fields, example:


```go
// Counter is just a silly example, never write a counter like this.
type Counter struct {
	sync.Mutex
	count int64
}

func (c *Counter) Up() {
	// Call promoted method from sync.Mutex.
	// Equivalent to c.Mutex.Lock()
	c.Lock()
	defer c.Unlock()
	c.count++
}
```

> > After the change, we're potentially breaking users by changing the public exported API of the package, since now DefaultCleaner has no exported fields whatsoever.
> 
> Yes, that was my intention because we're using just 2 and "promoting" 30.

That's why IMHO it's better to _default to unexported_, and only export things that have reason to be exported. That gives us freedom to refactor without breaking APIs.